### DISCUSSION TITLE
Start XState inspector locally

### DISCUSSION THREAD

**Discussion by evangeloszotos** [statelyai/xstate#1626] [Q&A]
Hi,

xstate-inspect works great for webapps.

I tried it in a react app.

I would like to run it on a react native app though.

An inspector UI is available at "https://statecharts.io/inspect" where my app connects to.

Though, xstate-inspect tries to open a new window, or control an iframe.

I would like to run the server, that is hosted under "https://statecharts.io/inspect" locally. 
This way, my virtual/physical device would just need to connect to my local development server.

xstate-inspect would just need to connect to an already open browser tab, instead of opening a new one.

Is this possible at the moment? 

I was not able to find the that is running under "https://statecharts.io/inspect"

Of course some adjustments need to be made, but I think that would be really powerful.

Bests
Evangelos

**Comment by JayNaire** [comment]
I'm adding support for this question. Adding a local inspector UI feature is key to my embedded app which has no runtime connection to the web; I wanted a (modified) UI to save a lot of work getting the machine to visually say what it is doing and why.  Unless I've got the wrong end of the stick completely, that will not be possible at the moment without a lot of effort replicating "https://statecharts.io/inspect.

**Reply by davidkpiano** [reply]
@JayNaire It's possible to create your own inspector locally now. See [the docs](https://xstate.js.org/docs/packages/xstate-inspect/#implementing) for this.

**Reply by JayNaire** [reply]
Thanks David; I have toyed fitfully with xstate-inspect but have not found success yet due to my lack of experience and frequent power outages :(  I shall persevere!

**Comment by m-deacon** [comment]
@evangeloszotos @JayNaire were you able to get the inspector working in your RN project? Could you provide some guidance for others (myself) trying to do this please? Otherwise, @davidkpiano is there perhaps a guide for setting this up? The docs for implementing your own inspector were a little confusing for me.

**Reply by JayNaire** [reply]
@evangeloszotos 
_were you able to get the inspector working in your RN project?_ 
Yes; very successfully but, to be clear, my app wasn't using React at all. Only the inspector.
All I did was to save https://statecharts.io/inspect locally (eg; Save page complete in browser) to a local folder and then dish that out as public statics (via Express in my case). That part worked easily, it's only necessary to point the browser or iframe to https://localhost:xxxx/inspect once the app's running (and you've got the routing sorted out).
My notes:
1. If you have multiple xstate machines: It's best not to splash ```service = interpret(m, { devTools: true })``` all over the place (inspect slows down a lot due, in my case, to Linux running into swap fairly quickly due to websocket activity).
2. There's a note somewhere to the effect of ```import { inspect  } from "@xstate/inspect/lib/server.js";``` before instantiating machines you want to inspect. I wasted ages not realising my code effectively wasn't doing this!
3. Leaving a browser's debugger (Chrome or Firefox) open on the inspector page eats/leaks memory. (Might be a generic problem  for me; not specifically inspect).
